src/c4_client/p1_data_structure.rs

```diff
@@ -0,0 +1,106 @@
+//! Before we implement any serious  methods on our client, we must, create the `Block` and `Header`
+//! data structures one last time like we did in Chapter 1. the logic you wrote there
+//! will be useful here as well and can probably be reused to some extent.
+//!
+//! Throughout the Blockchain chapter, we created a blockchain data structure that had:
+//! 1. a built-in addition accumulator state machine
+//! 2. A built-in pow consensus mechanism
+//!
+//! In the State Machine and Consensus chapters, we designed abstractions over both
+//! the state machine and the consensus. We also implemented several examples of each
+//! trait.
+//!
+//! This will be the last time we have to write this blockchain data structure,
+//! because this time it will be fully generic over both the state machine and consensus
+//! logic, thanks to our traits.
+//!
+//! This abstraction is the key idea behind blockchain _frameworks_ like Substrate or the Cosmos SDK.
+
+use super::{Consensus, ForkChoice, Header, StateMachine};
+
+use super::FullClient;
+type Hash = u64;
+
+impl<Digest> Header<Digest> {
+    /// Returns a new valid genesis header.
+    fn genesis(genesis_state_root: Hash) -> Self {
+        todo!("Exercise 1")
+    }
+
+    /// Create and return a valid child header.
+    fn child(&self, state_root: Hash, extrinsics_root: Hash) -> Self {
+        todo!("Exercise 2")
+    }
+
+    /// Verify a single child header.
+    fn verify_child(&self, child: &Self) -> bool {
+        todo!("Exercise 3")
+    }
+
+    /// Verify that all the given headers form a valid chain from this header to the tip.
+    fn verify_sub_chain(&self, chain: &[Self]) -> bool {
+        todo!("Exercise 4")
+    }
+}
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Block<C: Consensus, SM: StateMachine> {
+    header: Header<C::Digest>,
+    body: Vec<SM::Transition>,
+}
+
+impl<C: Consensus, SM: StateMachine> Block<C, SM> {
+    /// Returns a new valid genesis block. By convention this block has no extrinsics.
+    pub fn genesis(genesis_state: &SM::State) -> Self {
+        todo!("Exercise 5")
+    }
+
+    /// Create and return a valid child block.
+    pub fn child(&self, pre_state: &SM::State, extrinsics: Vec<u8>) -> Self {
```

Comment:
a bit confused on the pre_state term but i guess I get it.
 
 So you basically just hash the `extrinsics` and then the `pre_state` and fill the header with the rest of the fields and place the extrinsics as the `body` field of the `Block` struct?
 
 I guess I got a little confused on pre_state.

Reply:
For the extrinsics, yeah, you just hash them and put the hash in the header.

But for the state it is different. You are given a starting state (which is also the ending state of the parent block). Then you execute the transactions to get the new ending state. Then you hash that new ending state and put it in the header.